### PR TITLE
Feat/repay close ladle module

### PR DESCRIPTION
**Pull request by Sabnock01**

### CODE REVIEW

File: packages/foundry/contracts/test/modules/RepayCloseModule.t.sol [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    function testRepayFromLadle() public {
+        console.log("Can repay from ladle");
+        uint256 joinBalanceBefore = dai.balanceOf(address(join));
+        uint256 baseBalanceBefore = dai.balanceOf(address(this));
+        vm.prank(address(ladle));
+        ILadleCustom(address(ladle)).moduleCall(
+            address(module),
+            abi.encodeWithSelector(module.repayFromLadle.selector, vaultId, address(this), address(this))
```
Comment (alcueca) [inline packages/foundry/contracts/test/modules/RepayCloseModule.t.sol:63]: Please test with different receivers for the collateral and the remainder. Same below.

**Review by alcueca — CHANGES_REQUESTED** [review]
Looks good, but the tests need to test the main feature change :)

File: packages/foundry/contracts/modules/RepayCloseModule.sol [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    using CastU256I128 for uint256;
+    using CastU256U128 for uint256;
+    using TransferHelper for IERC20;
+
+    ///@notice Creates a Healer module for the Ladle
+    ///@param cauldron_ address of the Cauldron
+    ///@param weth_ address of WETH
+    constructor(ICauldron cauldron_, IWETH9 weth_) LadleStorage(cauldron_, weth_) {}
+
+    /// @dev Use fyToken in the Ladle to repay debt. Return unused fyToken to `to`.
```
Comment (alcueca) [inline packages/foundry/contracts/modules/RepayCloseModule.sol:19]: "Return collateral to foo and unused fyToken to bar"

File: packages/foundry/contracts/modules/RepayCloseModule.sol [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            IJoin ilkJoin = getJoin(vault.ilkId);
+            ilkJoin.exit(collateralReceiver, repaid.u128());
+        }
+
+        // Return remainder
+        if (amount - repaid > 0) IERC20(address(series.fyToken)).safeTransfer(remainderReceiver, amount - repaid);
+
+    }
+
+    /// @dev Use base in the Ladle to repay debt. Return unused base to `to`.
```
Comment (alcueca) [inline packages/foundry/contracts/modules/RepayCloseModule.sol:47]: Same as above

File: packages/foundry/contracts/modules/RepayCloseModule.sol [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    /// Return as much collateral as debt was repaid, as well. This function is only used when
+    /// removing liquidity added with "Borrow and Pool", so it's safe to assume the exchange rate
+    /// is 1:1. If used in other contexts, it might revert, which is fine.
+    function closeFromLadle(bytes12 vaultId_, address collateralReceiver, address remainderReceiver)
+        external payable
+        returns (uint256 repaid)
+    {
+        (bytes12 vaultId, DataTypes.Vault memory vault) = getVault(vaultId_);
+        DataTypes.Series memory series = getSeries(vault.seriesId);
```
Comment (alcueca) [inline packages/foundry/contracts/modules/RepayCloseModule.sol:56]: You should actually check that `series.fyToken.underlying() == cauldron.assets(vault.ilkId)`, or some other cheaper method to check that this is only used for repaying "Borrow and Pool" vaults where the series underlying and the collateral are the same.

File: packages/foundry/contracts/modules/RepayCloseModule.sol (hunk 2 above)
Comment (Sabnock01) [reply packages/foundry/contracts/modules/RepayCloseModule.sol:19]: Of course. Don't know why I neglected that part.

File: packages/foundry/contracts/test/modules/RepayCloseModule.t.sol (hunk 1 above)
Comment (Sabnock01) [reply packages/foundry/contracts/test/modules/RepayCloseModule.t.sol:63]: Of course. Don't know why I neglected that part.

File: packages/foundry/contracts/test/modules/RepayCloseModule.t.sol [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        vm.prank(address(ladle));
+        bytes memory data = ILadleCustom(address(ladle)).moduleCall(
+            address(module),
+            abi.encodeWithSelector(module.repayFromLadle.selector, vaultId, foo, bar)
+        );
+        uint256 repaid = abi.decode(data, (uint256));
+
+        assertEq(repaid, WAD);
+        assertEq(joinBalanceBefore, dai.balanceOf(address(join)) + WAD);
+        assertEq(baseBalanceBefore, dai.balanceOf(address(foo)) - WAD);
```
Comment (alcueca) [inline packages/foundry/contracts/test/modules/RepayCloseModule.t.sol:101]: You are still not testing that the unused fyToken go to `bar`. To do so, you just need to deal more fyDai to the Ladle than debt is there in the vault we are repaying.

File: packages/foundry/contracts/test/modules/RepayCloseModule.t.sol [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        vm.prank(address(ladle));
+        bytes memory data = ILadleCustom(address(ladle)).moduleCall(
+            address(module),
+            abi.encodeWithSelector(module.closeFromLadle.selector, vaultId, foo, bar)
+        );
+        uint256 repaid = abi.decode(data, (uint256));
+
+        assertEq(repaid, WAD);
+        assertEq(joinBalanceBefore, dai.balanceOf(address(join)));
+        assertEq(baseBalanceBefore, dai.balanceOf(address(foo)) - WAD);
```
Comment (alcueca) [inline packages/foundry/contracts/test/modules/RepayCloseModule.t.sol:120]: You are still not testing that the unused dai go to `bar`. To do so, you just need to deal more Dai to the Ladle than debt is there in the vault we are repaying.

**Review by alcueca — CHANGES_REQUESTED** [review]
The tests still need more coverage.

File: packages/foundry/contracts/test/modules/RepayCloseModule.t.sol (hunk 5 above)
Comment (Sabnock01) [reply packages/foundry/contracts/test/modules/RepayCloseModule.t.sol:101]: There was some confusion about this but I see now.

**Review by alcueca — APPROVED** [review]
Well done!

**Pull request merged**